### PR TITLE
DRAFT: vmalert add multi-tenant support

### PR DESCRIPTION
**Pull request by dxtrzhang**
mvp version of multi-tenant support, see #740

### CODE REVIEW

**Comment by hagen1778** [comment]
Would it make sense to just override datasource URL in group config?
```
groups:
- name: "alert name"
  datasource: "http://localhost:8428/select/1/prometheus/2"
  rules:
  - ...
  - ...
`
``
This should simplify the code and increase flexibility since you can configure different URLs now. On other hand, configuration becomes more verbose.

**Comment by dxtrzhang** [comment]
> Would it make sense to just override datasource URL in group config?

If we use `datasource` field, it means vmalert can support multi-datasource, this will change the way how vmalert dealing with the http client in `VMStorage`.  
The reason why I parseURL in code is that I don't want to break the current version's config flags, otherwise, we can set a `baseURL`, a `suffix`,  a `defaultAuthToken`, it can also simplify the code and increase flexibility.

PS: Multi-datasource support is an interesting feature too, but not that necessary to me(the number of datasource is always controllable).

File: app/vmalert/datasource/vm.go [hunk 1]
```diff
@@ -49,28 +51,115 @@ const queryPath = "/api/v1/query?query="
...
-	basicAuthPass string
+	c                *http.Client
+	baseURL          string
+	suffix           string
+	basicAuthUser    string
+	basicAuthPass    string
+	defaultAuthToken *auth.Token
+}
+
+func parseURL(s string) (string, string, *auth.Token, error) {
```
Comment (hagen1778) [inline app/vmalert/datasource/vm.go:62]: Can we reuse `httpserver.ParsePath()` here?

**Review by hagen1778 — COMMENTED** [review]
I wonder if it makes sense to define multiple tenants for the same group? For example, there are 10 tenants and all of them rely on the alerts/recordings for node-exporter metrics. In this case, user will need to copy the same groups 10 times. Instead, he/she could just list tenants in one group. On the other hand, this will make the code more complex since now one group will need to query all the tenants concurrently. wdyt?

File: app/vmalert/datasource/vm.go (hunk 1 above)
Comment (dxtrzhang) [reply app/vmalert/datasource/vm.go:62]: Thought about using `httpserver.ParsePath()`, but this function will raise error if the tenancy was not set.    

Actually I don't like the way that we have to parse the input URL. I am wondering what if we dump `datasource.url` or delete tenancy parts: 
1. add a global tenancy setting maybe like:
    ```yml
    global:
      tenant:
        accountID: 42
        projectID: 12
    groups:
    - name: ...
    ```
2. some command line flags like `tenant.globalAccountID` and `tenant.globalProjectID`.

What do you think? Will this be unfriendly to users?

PS: prefer the first one.

**Comment by dxtrzhang** [comment]
> I wonder if it makes sense to define multiple tenants for the same group

For me, I think It is not worth to complex the code logic.   

First, tenants should not aware each others' existence. Second, I don't think users actually "write" the rule file themselves, what users actually write are alert/recording rules, and then use some tools to generate the rule file, which is OK for the repeat, and I think the repeat happens to make the rule file more human readable :)

File: app/vmalert/datasource/vm.go (hunk 1 above)
Comment (hagen1778) [reply app/vmalert/datasource/vm.go:62]: Not sure if I understand. I assume there should two options to configure vmalert:
1. Configure vmalert per tenant. User specifies full `datasource.url` with auth token. That's how many people do now and we should should keep compatibility.
2. Configure multiple tenants for one vmalert. User specifies `datasource.url` to point the `vmselect` address only and omits auth token. Then user specifies tenant account and project IDs per-group or per-file (global config from p.1) and vmalert builds the correct URL for each group.
This looks natural and shouldn't break existing installations.

**Comment by hagen1778** [comment]
One more thing to consider - group uniqueness. For now, group's unique ID consists of hash for file name and group name. Probably, tenant configuration should be taken into consideration if we allow to configure tenant per-group. Could be also nice to attach tenants config as labels to generated alerts and recording rules.

**Comment by dxtrzhang** [comment]
> Could be also nice to attach tenants config as labels to generated alerts and recording rules.

Adding tenants config as labels to generated alerts looks good to me, since alertmanager should see alerts from two tenants as different alerts.   
About recording rules, why not save data under the associated tenants of the configured remote write.

File: app/vmalert/datasource/vm.go (hunk 1 above)
Comment (dxtrzhang) [reply app/vmalert/datasource/vm.go:62]: So these two options should both be available for vmalert? First one for capacity and second one for long term using, Right?

**Comment by hagen1778** [comment]
> About recording rules, why not save data under the associated tenants of the configured remote write.

Yes, this sounds good and logical.

**Comment by dxtrzhang** [comment]
Some configuration issues for vmalert:
1. current flag `-rule` supports patterns, and it is the rule path config, maybe not suitable for configs like `global` we discuss above, what about letting vmalert supports config file & flags
2. the priorities between config file and flag, lets's say, if I set `-datasource.url=http://127.0.0.1:8428/select/42:12` and set tenant in config file, which one should be overwritten?

**Comment by hagen1778** [comment]
> about global, is it instance global or rule file global

My understanding was that it is global-per-file - means only groups and rules listed in the file will be affected by the global config. 

> what about letting vmalert supports config file & flags

There are attempts to implement this - see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/730

> the priorities between config file and flag, lets's say, if I set -datasource.url=http://127.0.0.1:8428/select/42:12 and set tenant in config file, which one should be overwritten?

Not sure either. Afaik there was discussion about global config file and implementation draft here https://github.com/VictoriaMetrics/VictoriaMetrics/pull/739. Thee description says "Configuration is based on flag names and overrides flag values". @valyala @f41gh7 can you provide some details about this?

**Comment by dxtrzhang** [comment]
> My understanding was that it is global-per-file - means only groups and rules listed in the file will be affected by the global config.

Looks good to me!   
If it is file global, instance global should also be provided though, may be via flag or config file, wdyt? Is this necessary to set instance global tenant?

**Comment by hagen1778** [comment]
> vmalert has three flags(datasource.url, remoteWrite.url and remoteRead.url), for now they are separately defined

Yeah, this make things complicated sometimes. However, we introduce a very new piece of configuration which could contain new configuration rules. Let's get back to our example:
```
global:
  tenant:
    accountID: 42
    projectID: 12
groups:
- name: ...
```

Let's assume that anything set in config file overrides the flag values. Hence, the backward compatibility will be kept because old installation don't have the new config. Based on config above we can assume that both `remoteWrite` and `remoteRead` URLs are equal to the `datasource.url`. We also can provide ability to specify all three URLs per-group or per-file:
```
global:
  datasource:
    url: http://localhost:8428
  remoteWrite:
    url: http://localhost:9428
  remoteRead:
    url: ""
  tenant:
    accountID: 42
    projectID: 12
groups:
- name: foo
    remoteRead:
      url: http://localhost:9428
```

Let's "parse" the config above:
1. All groups in the file inherit the `datasource.url` value of `http://localhost:8428`. The flag `datasource.url` is ignored.
2. All groups in the file inherit the `remoteWrite.url` value of `http://localhost:9428`. The flag is ignored
3. The `remoteRead` is disabled for all groups, except of group `foo`, which inherits behaviour from p.1 and p.2 but also overrides the `remoteRead.url` by its own value of ` http://localhost:9428`

If any of URL values omitted, the flag values should be used. If tenant is specified, remoteRead/Write enabled automatically. Potentially, we can change existing behaviour for the flags and automatically set remoteRead/Write flag values to `datasource.url` value.

The approach looks a bit complicated. We can start with global config only and skip the override mechanism. But the initial design should keep it in mind.

Since I'm not dogfooding `vmalert` myself I'd really appreciate your opinion regarding the convenience of this. Opinions from @valyala and @f41gh7 are also welcomed.

**Comment by dxtrzhang** [comment]
Thanks for the reply :)

Actually I don't want to add `datasource.url`, `remoteRead.url` or `remoteWrite.url` to rule files yet, since #739 are still under discussion, too many unstable factors.

---
What I do for now is getting the instance wide default tenant by parsing `datasource.url`, and if any group sets the tenant or file global tenant is set,  the instance wide default tenant will be overwritten for that group.  

Let's say that we have `datasource.url=https://127.0.0.1:8481/select/1/prometheus`, and rule files like:
```yml
# file1
global:
  tenant:
    accountID: 42
groups:
- name: group1
  ...
- name: group2
  tenant:
    projectID: 1
```
```yml
# file2
groups:
- name: group3
  ...
```
the parsed tenant configs are:
  * group1 -> 42:0
  * group2 -> 42:1
  * group3 -> 1:0

---
About `remoteRead.url` and `RemoteWrite.url`.  
1. If the `remoteRead.url` and `RemoteWrite.url` is not set, we enable the remoteWrite/remoteRead automatically, and save the data under the associated tenant.
2. If the `remoteRead.url` and `RemoteWrite.url` is set, we save the data to the configured url without any other settings since
    * former version of vmalert  has already saved some data under that remote storage, we should keep doing that to keep compatibility
    * the configured remote storage may be another vm instance, and we never know if it is a cluster version or single version.

`remoteWrite.url` and `remoteRead.url` are optional in the future in case that users may want to save rule generated data to somewhere else.

This option keeps compatibility and the user habit while including the multi-tenancy into vmalert. More aggressive configurations change can be done after #739 is merged or clearly discussed.

wdyt? @hagen1778 @valyala @f41gh7

**Comment by f41gh7** [comment]
I think, that explicit is better than implicit. If we define tenant settings at some groups, we must provide corresponding tenant configuration for remoteWrite, remoteRead, notifier and datasource. Otherwise, assume, that this setting is not set for that group. It protects us from configuration abusing, misconfiguration and security issues. If tenant is empty, use default configuration, if its defined.

 About configuration file, i think it's better to define separate  one file with tenant configuration. Merging files is painful. I think about something like that:
```# default only one url per group
datasources:
  - url: http://vmselect-1/basepath
    tenants:
      - 1:10
      - 2:30
  - url: http://vmselect-1/basepath
    tenants:
      - 1:15
      - 2:0
  - url: http://vmselect-3/basepath
    default: true
    tenants:
      - 3:0
  # its possible to use vm in non-cluster mode
  - url: http://vmsingle-database
    # disables url formating with projectID:accountID
    format_url: false
    tenants:
      - 5:0


notifier:
  - url: http://notifier-1/path
    tenants:
      - 1:10
      - 2:30
      - 2:0
      - 3:0
  - url: http://notifier-2
    default: true

remoteWrite:
  - url: http://vminsert-2/path
    default: true

remoteRead:
  - url: http://select-2/path
    default: true


#at rules file
groups:
  # maps to remotewrite, remoteread and datasource with 1:10 tenant
 - name: group2
   tenant:
     projectID: 1
     accountID: 10

  # maps to default tenant
 - name: group2

  # invalid group, no projectID
 - name: group2
   tenant:
     accountID: 10

  # maps to 1:0 tenant
 - name: group2
   tenant:
     projectID: 1

  # maps to 5:0 tenant vmsingle
 - name: group2
   tenant:
     projectID: 5```
  
 In theory, we can define it with flags: `--notifier.url=http://notifier-2`, `--notifier.default=true`, `--notifier.url=http://notifier-1/path`, `--notifier.tenants=1:10,2:30, 2:0,3:0`. But it looks like madness) 

 And i don't think, that we should support both flags and configuration file for urls (remotewrite,remoteread, notifier and datasource), it's too complicated. Only flags or configuration file at one time is possible.

**Comment by dxtrzhang** [comment]
> I think, that explicit is better than implicit

Totally agree with that @f41gh7 ! 

What I concern here is the back compatibility. Assume if users have set the `remoteWrite.url` and `remoteRead.url`, we do have to save the generated data into that configured remote storage.   

The reason why I agree with the idea of "automatically enable the remoteWrite/remoteRead and save the data under the associated tenant" from @hagen1778  is that I think the ability to read the state of alert even after restarting vamlert is necessary(which users should not invoke in). 

I thought about dropping `remoteRead.url` and `remoteRead.url` to reduce the  complexity, but first, we have to keep back compatibility, and maybe some users may want to save these datas somewhere else.

---
About the config file, I know merging config files is painful, and agree with the idea of a standalone config file instead of flags. But this change is too aggressive, we have to warn users about that in the release note. So what happens to #739 is very important, and all the components should make the changes at same release.

**Comment by dxtrzhang** [comment]
Seems like #739 was temporarily closed. Any updates here?

**Comment by hagen1778** [comment]
> Seems like #739 was temporarily closed. Any updates here?

That PR was closed since it has high complexity and requires the driver who actually uses `vmalert`. So far, in my opinion, that PR may result only into conflicts with current PR. My understanding is that we need to make step-by-step changes instead doing them in parallel. Pls let me know your thoughts.

> If the remoteRead.url and RemoteWrite.url are not set, we enable the remoteWrite/remoteRead automatically, and save the data under the associated tenant.

We also need a way to disable read/write. Let's say by setting empty string to them in group global config.

> If the remoteRead.url and RemoteWrite.url are set, we save the data to the configured url without any other settings since

My current understanding is that flags - are default values for all groups. But if group has an override `vmalert` should stick to it. 
So with multi-tenancy support we introduce a new piece of config and we know exactly that no existing users will be affected. My opinion is that using the same url for queries and writes is a good thing to do and we should do this if tenant defined. The reasoning is that default values should be fine for most of user, and most of them are just setting the same value to 3 different flags (datasource, remoteRead, remoteWrite). For the sake of compatibility we should allow to change or disable remoteRead/remoteWrite values explicitly per-group.

**Comment by dxtrzhang** [comment]
> For the sake of compatibility we should allow to change or disable remoteRead/remoteWrite values explicitly per-group.

"per group" may not be a good idea or not necessary since current version support remoteRead/remoteWrite at instance level.

### Compatibility
If the `remoteWrite` and `remoteRead` is set, we respect the configured value, otherwise, we use the same url for queries and writes(aka `datasource`).   

It's hard to "disable"  the remoteWrite/remoteRead though, because the default value of the config is already empty character, we don't know if the users set empty character to disable the feature or just leave for default.

**Comment by hagen1778** [comment]
> "per group" may not be a good idea or not necessary since current version support remoteRead/remoteWrite at instance level.

sounds reasonable!

> It's hard to "disable" the remoteWrite/remoteRead though, because the default value of the config is already empty character, we don't know if the users set empty character to disable the feature or just leave for default.

Flag value and value parsed from config are different values and could be interpreted differently. We also can use an object to configure url, so then it would be easy to distinguish empty and nil values - plz see example at https://github.com/VictoriaMetrics/VictoriaMetrics/pull/754#issuecomment-692349926

> What I do for now is getting the instance wide default tenant by parsing datasource.url, and if any group sets the tenant or file global tenant is set, the instance wide default tenant will be overwritten for that group.

Let's stick to this approach for now. We can also skip tenant-per-group and keep tenant-per-file for the sake of simplicity. I guess, the first iteration of multitenancy support in vmalert should do the following:
1. Solve your needs and needs described here https://github.com/VictoriaMetrics/VictoriaMetrics/issues/740
2. Allow to configure tenant-per-file, so users could generate multiple files instead of running multiple vmalerts
3. Support multitenancy for remoteWrite/remoteRead

I think we need to make it as simple as possible for now, and then iterate with better/more flexible tenant support. wdyt @dxtrzhang ?
Plz feel free to reach me in VM slack if you think live discussion is needed.

**Comment by dxtrzhang** [comment]
> Flag value and value parsed from config are different values and could be interpreted differently. We also can use an object to configure url, so then it would be easy to distinguish empty and nil values - plz see example at #754 (comment)

According to my understanding, remoteWrite and remoteRead is a instance wide config which should not be set in rule file.

**Comment by hagen1778** [comment]
Closing due to inactivity. Please also see https://docs.victoriametrics.com/vmalert.html#multitenancy

**Pull request closed without merging**